### PR TITLE
Porting to `Filepath` class

### PR DESCRIPTION
**Pull request by bogdasar1985**
Issue https://github.com/newsboat/newsboat/issues/2326

Progress:
- [x]  cache.h
- [x]  cliargsparser.h
- [x]  colormanager.h
- [x]  configactionhandler.h
- [x]  configcontainer.h
- [x]  configdata.h
- [x]  configexception.h
- [x]  confighandlerexception.h
- [x]  configparser.h
- [x]  configpaths.h
- [x]  controller.h
- [x]  curldatareceiver.h
- [x]  curlhandle.h
- [x]  dbexception.h
- [x]  dialogsformaction.h
- [x]  dirbrowserformaction.h
- [x]  download.h
- [x]  emptyformaction.h
- [x]  exception.h
- [x]  feedcontainer.h
- [x]  feedhqapi.h
- [x]  feedhqurlreader.h
- [x]  feedlistformaction.h
- [x]  feedretriever.h
- [x]  filebrowserformaction.h
- [x]  filepath.h
- [x]  file_system.h
- [x]  fileurlreader.h
- [x]  filtercontainer.h
- [x]  fmtstrformatter.h
- [x]  formaction.h
- [x]  freshrssapi.h
- [x]  freshrssurlreader.h
- [x]  fslock.h
- [x]  helpformaction.h
- [x]  history.h
- [x]  htmlrenderer.h
- [x]  inoreaderapi.h
- [x]  inoreaderurlreader.h
- [x]  itemlistformaction.h
- [x]  itemrenderer.h
- [x]  itemutils.h
- [x]  itemviewformaction.h
- [x]  keymap.h
- [x]  links.h
- [x]  listformaction.h
- [x]  listformatter.h
- [x]  listmovementcontrol.h
- [x]  listwidgetbackend.h
- [x]  listwidget.h
- [x]  logger.h
- [x]  matchable.h
- [x]  matcherexception.h
- [x]  matcher.h
- [x]  minifluxapi.h
- [x]  minifluxurlreader.h
- [x]  newsblurapi.h
- [x]  newsblururlreader.h
- [x]  nullconfigactionhandler.h
- [x]  ocnewsapi.h
- [x]  ocnewsurlreader.h
- [x]  oldreaderapi.h
- [x]  oldreaderurlreader.h
- [x]  opml.h
- [x]  opmlurlreader.h
- [x]  pbcontroller.h
- [x]  pbview.h
- [x]  poddlthread.h
- [x]  queueloader.h
- [x]  queuemanager.h
- [x]  regexmanager.h
- [x]  regexowner.h
- [x]  reloader.h
- [x]  reloadthread.h
- [x]  remoteapi.h
- [x]  rssfeed.h
- [x]  rssignores.h
- [x]  rssitem.h
- [x]  rssparser.h
- [x]  ruststring.h
- [x]  scopemeasure.h
- [x]  searchresultslistformaction.h
- [x]  selectformaction.h
- [x]  statusline.h
- [x]  stflpp.h
- [x]  strprintf.h
- [x]  tagsouppullparser.h
- [x]  textformatter.h
- [x]  textviewwidget.h
- [x]  ttrssapi.h
- [x]  ttrssurlreader.h
- [x]  urlreader.h
- [x]  urlviewformaction.h
- [x]  utils.h
- [x]  view.h
-----------
- [x] RSS
- [x] FILTER

### CODE REVIEW

File: test/opml.cpp
Comment (Minoru) [inline test/opml.cpp:305]: `std::string()` is a temporary kludge; don't use it explicitly. Use `.to_locale_string()` instead.

Ditto in all other places below except the ones where I mention `.display()`.

File: test/test_helpers/chdir.cpp
Comment (Minoru) [inline test/test_helpers/chdir.cpp:16]: This message is going to be displayed to the user, so we should use `.display()` here to get a valid UTF-8 string.

File: test/test_helpers/chmod.cpp
Comment (Minoru) [inline test/test_helpers/chmod.cpp:15]: `.display()` (see rationale above).

**Review by Minoru — CHANGES_REQUESTED** [review]
Good start! Apart from the comments below, please work on two more things:

1. rebase these changes onto the target branch. This will remove the ≈100 unrelated commits and make this easier to review (and our Git history cleaner)
2. can you please take a look at CI and figure out why it fails to compile anything? I know it was broken before your changes, but it's quite unnerving to develop anything when CI is red :) There's a lot of changes before yours, so it's okay if you can't get to the bottom of it; just share what you've found and we'll figure it out together.

**Comment by bogdasar1985** [comment]
@Minoru 
https://github.com/newsboat/newsboat/blob/fe6eedd0a3caf090a7228af772881b4fb412ae91/include/filepath.h#L85C1-L89C44
```cpp
	Filepath(const Filepath&) = delete;
	Filepath& operator=(const Filepath&) = delete;
	Filepath(Filepath&&) = default;
	Filepath& operator=(Filepath&&) = default;
```
Why copy constructor and operator marked as deleted? What the reason for it? It makes porting to `Filepath` difficult.
P.S. I confess, I am not good enough with Git, I will try to deal with this `rebase` soon))

File: test/configpaths.cpp
Comment (Minoru) [inline test/configpaths.cpp:17]: Ah, I see these functions are copied all over the place. Let's move them to test_helpers/misc.{h,cpp} instead.

File: src/feedhqurlreader.cpp
Comment (Minoru) [inline src/feedhqurlreader.cpp:15]: Why not `file(url_file.clone())`?

**Review by Minoru — COMMENTED** [review]
Good work! `ConfigPaths` and `CliArgParser` need more work because they return `String` on the Rust side and wrap that into `rust::String` on C++ side. Take a look at how I handled it here: https://github.com/newsboat/newsboat/compare/master...feature/2326-filepath#diff-39127ef91e8d635d432796ac85b4a66cf3dd408b8b4510256b7ea72975971812R25

**Comment by Minoru** [comment]
`Filepath` can't be copied because it contains `rust::Box`, which is move-only (it's basically equivalent to `std::unique_ptr`). I don't mind writing copy c-tor and operator in terms of `clone()` though. Path handling probably isn't the performance-sensitive area of Newsboat anyway.

File: src/configparser.cpp
Comment (Minoru) [inline src/configparser.cpp:108]: Please revert this change, because `location` is not a path to a file — it's a part of the error message for the user.

**Review by Minoru — CHANGES_REQUESTED** [review]

**Comment by Minoru** [comment]
> P.S. I confess, I am not good enough with Git, I will try to deal with this rebase soon))

`git checkout filestring && git rebase --onto feature/2326-filepath 0b4b432` should do the trick: it takes commits that you made after the merge with `master` (that's 0b4b432) and puts them on top of feature/2326-filepath. But I didn't test it :)

Alternatively, you can do the same thing manually:

1. `git checkout --branch filestring-2 feature/2326-filepath` -- this creates a temporary branch where you'll create a new history. That branch starts at feature/2326-filepath;
2. for each commit made after the merge with `master`: `git cherry-pick COMMIT`, e.g. `git cherry-pick e908576`;
3. check that the result compiles and passes the tests;
4. `git checkout filestring` to go back to your original branch;
5. `git reset --hard filestring-2` to make `filestring` equivalent to `filestring-2`;
6. `git push --force` to update this PR.

No matter what approach you take, you might run into some nasty merge commits, because it's been a while since I worked on feature/2326-filepath. If they are too hard to resolve, let's wrap up this PR (address all my current comments without porting any more classes to `Filepath`) and then I'll rebase the entire thing onto `master`. But I still think it's worthwhile for you to try doing it yourself, because that's pretty much he only way to learn to do it ;)

File: src/pbcontroller.cpp
Comment (Minoru) [inline src/pbcontroller.cpp:58]: `join` doesn't modify the object it's called on. Please use `push` here and further down this function.

**Review by Minoru — CHANGES_REQUESTED** [review]

File: test/view.cpp
Comment (Minoru) [inline test/view.cpp:33]: Instead of converting filepaths to strings and comparing them as strings, it's better to keep them as `Filepath`s and compare with `operator==`:

```c++
REQUIRE(
    v.get_filename_suggestion(example_en)
    ==
    Filepath::from_locale_string("Comparing_Visual.txt"));
```

**Review by Minoru — CHANGES_REQUESTED** [review]

File: src/itemlistformaction.cpp
Comment (Minoru) [inline src/itemlistformaction.cpp:1503]: Empty path is a kludge, let's get rid of it :)

1. on `master`, make `run_dirbrowser()` return `nonstd::optional<std::string>`;
2. wait for that change to get merged;
3. rebase this branch onto `master` and while doing so, update the code here to check if the `optional` is empty.

File: src/itemlistformaction.cpp
Comment (Minoru) [inline src/itemlistformaction.cpp:1517]: Why do we need `clone()` here? `join()` should return a new `Filepath` already.

File: src/itemlistformaction.cpp
Comment (Minoru) [inline src/itemlistformaction.cpp:1538]: Same question: why do we need `clone()` here?

File: include/download.h
Comment (Minoru) [inline include/download.h:29]: Why are these needed?

File: src/download.cpp
Comment (Minoru) [inline src/download.cpp:68]: Please add a `file_name` method to `Filepath` that calls this Rust method: https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.file_name Then use it here.

File: src/queueloader.cpp
Comment (Minoru) [inline src/queueloader.cpp:246]: Why do we need `clone()` here?

File: src/controller.cpp
Comment (Minoru) [inline src/controller.cpp:949]: Matching on the first byte of UTF-8 encoding is meh, although it should work fine. I see a couple options:

1. add this method to `Filepath` and rewrite the code to use it: https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.starts_with
2. add `is_absolute()` and `is_in_homedir()` methods to `Filepath` and rewrite the code with them. `is_absolute()` can defer to the same method on `PathBuf`, while `is_in_homedir()` you can implement on the Rust side with the aforementioned `starts_with()`.

File: src/controller.cpp
Comment (Minoru) [inline src/controller.cpp:947]: Why do we need a `clone()` here?

File: include/dirbrowserformaction.h
Comment (Minoru) [inline include/dirbrowserformaction.h:48]: Is this unused? If so, simply remove it.

File: src/dirbrowserformaction.cpp
Comment (Minoru) [inline src/dirbrowserformaction.cpp:343]: Please add https://doc.rust-lang.org/nightly/std/path/struct.PathBuf.html#method.set_extension to `Filepath` and rewrite this line to use that method instead.

File: src/filebrowserformaction.cpp
Comment (Minoru) [inline src/filebrowserformaction.cpp:322]: Please mark this with `TODO: #2326 use length by graphemes` and a link to this comment, because this place needs more work:

- need to check what units STFL uses for position: bytes of UTF-8, Unicode characters, graphemes?
- need to provide the length in appropriate units here

**Review by Minoru — CHANGES_REQUESTED** [review]
You're now touching algorithms that inspect the paths, so there are a lot more comments from me. The rule of thumb is to defer everything to Rust: it knows how to handle that stuff.

Each of those comments can take quite a bit of effort to resolve, so I'm also okay with a different approach: mark each place with a `TODO` pointing to the original issue and my comment, and wrap up this PR. You've already made a stellar progress here, so you decide how you'd like to proceed.

File: src/queueloader.cpp
Comment (bogdasar1985) [reply src/queueloader.cpp:246]: Without `clone()` we got this. I'm not going into `Filepath` class yet. I tried create copy ctor/optor, but run into problems (compile-errors), later I will try to solve it.
```bash
src/queueloader.cpp:246:41: error: use of deleted function ‘newsboat::Filepath::Filepath(const newsboat::Filepath&)’
  246 |   const Filepath filename = dl.filename();
      |                                         ^
In file included from include/download.h:7,
                 from include/queueloader.h:10,
                 from src/queueloader.cpp:1:
include/filepath.h:89:2: note: declared here
   89 |  Filepath(const Filepath&) = delete;
      |  ^~~~~~~~
```

File: src/itemlistformaction.cpp
Comment (bogdasar1985) [reply src/itemlistformaction.cpp]: Line 1537 not compiled without clone. Line 1538 I fixed.

File: include/download.h
Comment (bogdasar1985) [reply include/download.h:29]: Because after migrating from `std::string` to `Filepath` class `Download` now got a member of type `Filepath`.

https://github.com/newsboat/newsboat/blob/201a939e8a33e7f56df9e34f94737bce8e04ad8f/include/download.h#L61

Default copy ctor/optor would call a `Filepath` default ctor/optor, but they were marked as deleted. As consequence compile error.

Comment (Minoru) [reply include/download.h:29]: Please don't mark comments resolved. I think it's *my* right to decide if my comment was addressed or not ;)

I understand the rationale now, but I think the solution is quite brittle. If we ever add another field to `Download`, we'd have to remember to update these constructor and operator too. Chances are that we won't, and then boom!—we have a bug.

I'd prefer to add a copy constructor and operator to `Filepath` itself. That would also spare us from all these `clone()` calls which look fine in Rust but are quite outlandish in C++. What do you think?

File: src/queueloader.cpp
Comment (Minoru) [reply src/queueloader.cpp:246]: I think this can be fixed by changing the signature of `Download::filename()` from:

```c++
const newsboat::Filename filename() const;
```

to 


```c++
newsboat::Filename filename() const;
```

That way C++ can just move the result.

Ditto for `dirname()`.

**Comment by Minoru** [comment]
Good progress! You're getting into the weeds of Rust-C++ interoperability now, so please take it slowly. Commit early, commit often, and if you get stuck, just share the code and let's see what's up.

File: include/download.h
Comment (bogdasar1985) [reply include/download.h:29]: > Please don't mark comments resolved. I think it's my right to decide if my comment was addressed or not ;)
Ok, since now I will be leave a commit hash in such cases.

File: src/filebrowserformaction.cpp
Comment (bogdasar1985) [reply src/filebrowserformaction.cpp:323]: https://github.com/newsboat/newsboat/pull/2561/commits/e8ec713bc14b28b482c997cba650e9881199ea91

File: src/queueloader.cpp
Comment (bogdasar1985) [reply src/queueloader.cpp]: > Ditto for `dirname()`.

You probably meant `basename()`

https://github.com/newsboat/newsboat/pull/2561/commits/9e0aea3c2bc826e5e807ce1656d12b305dfd0004

File: include/download.h
Comment (Minoru) [reply include/download.h:29]: `Box` is indeed uncopiable, but you can ask Rust to clone it—see how `Filepath::clone()` is implemented. I think you can just copy the code from there and it will work. Once you got a copy constructor, `clone()` can be simplified to something like `return Filepath(*this)`. Can you please take another stab at this? Please push it even if it doesn't work, we'll figure it out.

**Comment by Minoru** [comment]
Latest commits look good!

File: include/download.h
Comment (bogdasar1985) [reply include/download.h]: Done! https://github.com/newsboat/newsboat/pull/2561/commits/aef5955cf2d373b8e5e17e4a768ed7a5d73dc156

File: src/dirbrowserformaction.cpp
Comment (bogdasar1985) [reply src/dirbrowserformaction.cpp]: https://github.com/newsboat/newsboat/pull/2561/commits/518a6a395021e6d62051f86352b6db6d9bf6f35c

File: src/controller.cpp
Comment (bogdasar1985) [reply src/controller.cpp]: https://github.com/newsboat/newsboat/pull/2561/commits/34cc025b8d3e3180c4f26d12e63f651ac6515588

File: src/download.cpp
Comment (bogdasar1985) [reply src/download.cpp]: https://github.com/newsboat/newsboat/pull/2561/commits/0e8adedc27881cacd0e25998cba7b91177c20775

File: include/download.h
Comment (Minoru) [reply include/download.h]: Excellent!

File: src/dirbrowserformaction.cpp
Comment (Minoru) [reply src/dirbrowserformaction.cpp]: Thank you for adding the test as well!

File: include/filepath.h
Comment (Minoru) [inline include/filepath.h:109]: This should return a `Filepath`, not a `string`, because the return value is a filename. This will also help us avoid all those `unwrap()` calls in the implementation of this method.

Also it should be wrapped in `nonstd::optional` to better match Rust's definition: https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.file_name

**Review by Minoru — CHANGES_REQUESTED** [review]

File: include/filepath.h
Comment (bogdasar1985) [reply include/filepath.h]: https://github.com/newsboat/newsboat/pull/2561/commits/58abe2f58b35ac722a485359ecee475b39608078

Rust type `Option<T>` is unsupported by `cxxbridge`, so I returned usual `String` and proceed empty string as `None`/`nullopt`

**Comment by bogdasar1985** [comment]
@Minoru can we get rid of `Filepath::clone()` method, cause of copy ctor/optor implemented.

File: include/filepath.h
Comment (Minoru) [reply include/filepath.h]: `String` uses UTF-8, so we can't use it for filenames. Please make it return `Vec<u8>` instead.

I'd also advise to use a separate `bool` out parameter to indicate `None`, so that we could return `Some("")`, but looking at Rust docs, apparently `file_name` returns either `None` or a non-empty string. So I think it's okay to rely on an empty vector in this case.

**Review by Minoru — COMMENTED** [review]
Looks better with every push! :)

**Comment by Minoru** [comment]
> maybe we can get rid of Filepath::clone() method, cause of copy ctor/optor implemented?

Yeah, it's a good idea I think. These `.clone()` calls look fine in Rust, but kind of alien in C++.

How do you want to approach the remaining comments? Would you like to pick them one by one, or would you prefer to create separate issues for them (to be resolved later) so this PR can be merged?

**Comment by bogdasar1985** [comment]
@Minoru 

> How do you want to approach the remaining comments? Would you like to pick them one by one, or would you prefer to create separate issues for them (to be resolved later) so this PR can be merged?

Only this https://github.com/newsboat/newsboat/pull/2561#discussion_r1357348873 as a separate PR.

Others I will do here.

File: include/filepath.h
Comment (bogdasar1985) [reply include/filepath.h]: https://github.com/newsboat/newsboat/pull/2561/commits/6338607602742ecf350661d14ae7b6d3d7dc62cf

File: src/itemlistformaction.cpp
Comment (bogdasar1985) [reply src/itemlistformaction.cpp:1503]: https://github.com/newsboat/newsboat/pull/2593

File: include/filepath.h
Comment (Minoru) [reply include/filepath.h]: Sorry for changing my mind so often, but since `Vec<u8>` doesn't save us from all those `unwrap()` calls, how about returning `Filepath` from the Rust side? It can be constructed from `OsStr` with `Path::new(str).to_owned()` (I think).

**Comment by Minoru** [comment]
The code looks much better without all those `clone()`s; thank you!

File: include/filepath.h
Comment (bogdasar1985) [reply include/filepath.h]: We can't use `Path`/`Box<Path>` as return type from Rust side, because it's fixed size array (here is inner implementation):
```rust
pub struct Slice {
    pub inner: [u8],
}
```

which leads to compile-time error:
```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> rust/libnewsboat-ffi/src/filepath.rs:16:14
   |
16 |         type Path;
   |              ^^^^ doesn't have a size known at compile-time
   |
   = help: within `filepath::Path`, the trait `Sized` is not implemented for `[u8]`
```
I use `PathBuf` instead.

File: rust/libnewsboat-ffi/src/filepath.rs
Comment (Minoru) [inline rust/libnewsboat-ffi/src/filepath.rs]: Please use `if let Some(res) = filepath.0.file_name() { ... }` — this avoids the call to `unwrap()`.

File: src/filepath.cpp
Comment (Minoru) [inline src/filepath.cpp:77]: This asks the compiler to convert `Box<PathBuf>` into `Filepath` and then convert *that* into `nonstd::optional<Filepath>`. I think you can help older compilers to chew through this with `return {str}` or maybe `return nonstd::optional<Filepath>(str)`.

File: include/filepath.h
Comment (Minoru) [reply include/filepath.h]: Using a `PathBuf` is the right call!

I think we can fix old compilers quite easily, see my comment below. If that won't work, sure, let's back out this change and write an issue so it can be fixed later.

File: src/filepath.cpp
Comment (bogdasar1985) [reply src/filepath.cpp:77]: It's not work.

`{str}` leads to
```bash
src/filepath.cpp: In member function ‘nonstd::optional_lite::optional<newsboat::Filepath> newsboat::Filepath::file_name() const’:
src/filepath.cpp:77:14: error: could not convert ‘{str}’ from ‘<brace-enclosed initializer list>’ to ‘nonstd::optional_lite::optional<newsboat::Filepath>’
   77 |   return {str};
      |              ^
      |              |
      |              <brace-enclosed initializer list>
```

`nonstd::optional<Filepath>(str)` can't be constructed

```bash
src/filepath.cpp: In member function ‘nonstd::optional_lite::optional<newsboat::Filepath> newsboat::Filepath::file_name() const’:
src/filepath.cpp:77:40: error: no matching function for call to ‘nonstd::optional_lite::optional<newsboat::Filepath>::optional(rust::cxxbridge1::Box<newsboat::filepath::bridged::PathBuf>&)’
   77 |   return nonstd::optional<Filepath>(str);
      |                                        ^
```

Let's revert commits.

**Comment by Minoru** [comment]
Okay, now we do two FFI trips (call Rust to get `Vec<u8>` which we then feed to Rust to get `Filepath`), but at least it works :) Can polish it later.

I believe the three remaining points are:

1. rebase onto `feature/2326-filepath`. Alternatively we can merge this PR as-is and I'll rebase everything onto `master` later
2. https://github.com/newsboat/newsboat/pull/2561#discussion_r1357348873 We can spin it off into a separate issue
3. https://github.com/newsboat/newsboat/pull/2561#pullrequestreview-1651417941 This too can become a separate issue

How would you like to proceed?

**Comment by bogdasar1985** [comment]
@Minoru let's 

> merge this PR as-is and _you'll_ rebase everything onto master later

**Comment by Minoru** [comment]
Okay! Thank you very much for the work. I'll merge and rebase it now, then create new issues and write an update on #2326.

**Review by Minoru — APPROVED** [review]

**Pull request merged**
